### PR TITLE
course learning materials sort and search additions

### PR DESCRIPTION
**Pull request by djvoa12**
**Summary:**
Course learning materials table is now FE sortable (with query params) and searchable.

**GIF:**

**Loading UI:**
Screen Shot 2019-03-25 at 1 48 57 PM

Accompanies https://github.com/ilios/frontend/pull/4450
Fixes https://github.com/ilios/frontend/issues/4402

### CODE REVIEW

File: addon/components/loading-spinner.js [hunk 1]
```diff
@@ -3,6 +3,10 @@ import layout from '../templates/components/loading-spinner';
 
 export default Component.extend({
   layout,
+
+  classNames: ['loading-spinner'],
   tagName: 'span',
-  classNames: ['loading-spinner']
+
+  class: '',
```
Comment (jrjohnson) [inline addon/components/loading-spinner.js]: This overloads this component too much and makes it harder to replace. Just stick with the regular spinner for now and don't worry about making it orange.

File: addon/mixins/course-materials-controller.js [hunk 2]
```diff
@@ -1,6 +1,8 @@
 import Mixin from '@ember/object/mixin';
 
 export default Mixin.create({
-  queryParams: ['sortBy'],
-  sortBy: 'firstOfferingDate',
+  queryParams: ['clmSortBy', 'slmSortBy'],
```
Comment (jrjohnson) [inline addon/mixins/course-materials-controller.js]: Names are a bit too obscure here. How about `courseSort` and `sessionSort`?

File: addon/mixins/sortable-table.js [hunk 3]
```diff
@@ -1,19 +1,21 @@
...
   sortBy: '',
 
   sortedAscending: computed('sortBy', function(){
     const sortBy = this.get('sortBy');
     return sortBy.search(/desc/) === -1;
   }),
+
   actions: {
-    sortBy(what){
+    sortBy(prop) {
```
Comment (jrjohnson) [inline addon/mixins/sortable-table.js]: Why make this change? It seems like unnecessary churn that adds to this diff. Should be a separate PR.

File: addon/templates/components/course-materials.hbs [hunk 4]
```diff
@@ -1,117 +1,174 @@
...
+
+  {{#if (gt (get (await this.course.learningMaterials) "length") 0)}}
+    <span class="filter-course-lms">
+      <input
+        placeholder={{t "general.filterPlaceholder"}}
+        value={{this.clmQuery}}
+        oninput={{perform this.setClmQuery value="target.value"}}>
+    </span>
+
+    <table class="table__course">
```
Comment (jrjohnson) [inline addon/templates/components/course-materials.hbs]: This class seems to only be used in tests. If it's for test selection you should use https://github.com/simplabs/ember-test-selectors which would look like `<table data-test-course-table>` .

File: addon/templates/components/course-materials.hbs [hunk 5]
```diff
@@ -1,117 +1,174 @@
...
-        value={{filter}}
-        oninput={{action (mut filter) value="target.value"}}
         placeholder={{t "general.filterPlaceholder"}}
-      >
+        value={{this.slmQuery}}
+        oninput={{perform this.setSlmQuery value="target.value"}}>
     </span>
-    <table>
+
+    <table class="table__session">
```
Comment (jrjohnson) [inline addon/templates/components/course-materials.hbs]: Should use test selectors here as well.

File: addon/components/course-materials.js [hunk 6]
```diff
@@ -1,127 +1,152 @@
...
-import SortableTable from 'ilios-common/mixins/sortable-table';
+import { or, reads } from '@ember/object/computed';
+import { isEmpty, isPresent } from '@ember/utils';
+import { all, task, timeout } from 'ember-concurrency';
 import layout from '../templates/components/course-materials';
 
-export default Component.extend(SortableTable, {
+const DEBOUNCE_DELAY = 250;
+
+export default Component.extend({
```
Comment (jrjohnson) [inline addon/components/course-materials.js]: This component is really hard to understand. It was already in pretty bad shape and adding all of these additional things to it made it much more difficult. I think it would be significantly simplified by using only computed properties. Can you spend some time trying that and we can pair on it as well.

**Comment by djvoa12** [comment]
@jrjohnson I addressed all your concerns with latest three commits.

File: addon/components/course-materials.js [hunk 7]
```diff
@@ -1,126 +1,156 @@
...
+  sessionSort: null,
+  slmQuery: '',
+  onClmSort() {},
+  onSlmSort() {},
+  typesWithUrl: Object.freeze(['file', 'link']),
+
+  isClmLoading: reads('courseLearningMaterialObjects.isPending'),
+  isSlmLoading: reads('sessionLearningMaterialObjects.isPending'),
+
+  courseLearningMaterialObjects: computed(function() {
```
Comment (jrjohnson) [inline addon/components/course-materials.js]: Instead of returning a PromiseArray this should be written as ```
courseLearningMaterialObjects: computed(async function() {
    return this.fetchCourseLearningMaterials();
  }),
```

I'm also not sure why `fetchCourseLearningMaterials` has been extracted. Especially since that obscures that there should be dependent keys on `course` and `course.sessions.[]` and probably other things. Why not just include all of those in a single computed property?

File: addon/components/course-materials.js [hunk 8]
```diff
@@ -1,126 +1,156 @@
...
-              });
-            });
-          });
-        }).then(lmObjects => {
-          resolve(lmObjects);
-        });
-      });
-    });
+
+  sessionLearningMaterialObjects: computed(function() {
```
Comment (jrjohnson) [inline addon/components/course-materials.js]: Same comments as course above

File: addon/components/course-materials.js [hunk 9]
```diff
@@ -1,126 +1,156 @@
...
-          let flat = learningMaterials.reduce((flattened, obj) => {
-            return flattened.pushObjects(obj.toArray());
-          }, []);
-
-          resolve(flat);
-        });
-      });
-    });
+
+  clmQuery: '',
```
Comment (jrjohnson) [inline addon/components/course-materials.js]: Since this is the external API of this component it would be clearer to define `clmQuery` and `slmQuery` as `courseQuery` in the same way as `courseSort`.

File: addon/components/course-materials.js [hunk 10]
```diff
@@ -1,126 +1,156 @@
...
+  course: null,
+  courseSort: null,
+  sessionSort: null,
+  slmQuery: '',
+  onClmSort() {},
```
Comment (jrjohnson) [inline addon/components/course-materials.js]: Am I missing where these two methods are used? What is their purpose?

File: addon/components/course-materials.js [hunk 11]
```diff
@@ -1,126 +1,156 @@
...
-                   (isPresent(obj.get('sessionTitle')) && obj.get('sessionTitle').match(exp));
-          });
-
-          resolve(filteredObjs);
-        }
-      });
-    });
+
+  filteredCourseLearningMaterialObjects: computed(
+    'courseLearningMaterialObjects', 'clmQuery', function() {
```
Comment (jrjohnson) [inline addon/components/course-materials.js]: `courseLearningMaterialObjects` returns a PromiseArray. I commented obove that it should return a promise. In either case this method should be async or depend on `courseLearningMaterialObjects.[]`.

File: addon/components/course-materials.js [hunk 12]
```diff
@@ -1,126 +1,156 @@
...
+      firstOfferingDate,
+      sessionTitle: session.title,
+      title: lm.title,
+      type: lm.type,
+      url: lm.url
+    });
+  },
+
+  filterClmo(clmo) {
+    const exp = new RegExp(this.clmQuery, 'gi');
```
Comment (jrjohnson) [inline addon/components/course-materials.js]: This query will need to be escaped before it can be used here. There is a cleanQuery util for this.  I also seem to recall that @stopfstedt was using a non-regex pattern for doing this kind of filtering and can probably point you in that direction instead.

File: addon/components/course-materials.js (hunk 9 above)
Comment (djvoa12) [reply addon/components/course-materials.js]: 👍

File: addon/components/course-materials.js (hunk 7 above)
Comment (djvoa12) [reply addon/components/course-materials.js]: I thought about that. How would you get a loading state going while those async functions run?

File: addon/components/course-materials.js (hunk 10 above)
Comment (djvoa12) [reply addon/components/course-materials.js]: The empty functions are used to signify functions are being passed in from parent component, just like a normal attribute. It's a common pattern.

Example:
```
hbs
{{child-component isLoading=isLoading onChange=(action "whatever")}}

js
isLoading: null,
onChange() {}
```

File: addon/components/course-materials.js (hunk 12 above)
Comment (djvoa12) [reply addon/components/course-materials.js]: I just used what I extracted from below without changing anything.

https://github.com/ilios/common/blob/master/addon/components/course-materials.js#L67

I can look into it further if you feel that more is needed.

File: addon/components/course-materials.js (hunk 7 above)
Comment (jrjohnson) [reply addon/components/course-materials.js]: You can use the [ember-promise-helpers](https://github.com/fivetanley/ember-promise-helpers) helpers for that.  In the template `{{#if (is-pending courseLearningMaterialObjects)}}` will give you a loading block and `{{#each (await courseLearningMaterialObjects) as |obj|}} will give you access to the promise results.

File: addon/components/course-materials.js (hunk 10 above)
Comment (jrjohnson) [reply addon/components/course-materials.js]: What I'm not seeing is where these are actually called in this component. Did the name change?

File: addon/components/course-materials.js (hunk 12 above)
Comment (jrjohnson) [reply addon/components/course-materials.js]: Yes, please change. Unescaped regex has caused us problems in the past if users enter a `/`. Pre-existing bug and not your fault, but worth cleaning up here.

File: addon/components/course-materials.js (hunk 10 above)
Comment (djvoa12) [reply addon/components/course-materials.js]: I changed them to `onCourseSort` and `onSessionSort`.

File: addon/templates/components/course-materials.hbs [hunk 13]
```diff
@@ -1,106 +1,153 @@
...
               {{/if}}
             </td>
-            <td>{{lmObject.type}}</td>
+            <td>{{capitalize lmObject.type}}</td>
             <td>{{lmObject.author}}</td>
           </tr>
+        {{else}}
+          <tr>
+            <td colspan="6" align="center">
+              {{if this.courseQuery "No Results Found" "No Course Learning Materials Available"}}
```
Comment (stopfstedt) [inline addon/templates/components/course-materials.hbs]: Please pull this verbiage out and into translations.

File: addon/templates/components/course-materials.hbs [hunk 14]
```diff
@@ -111,8 +158,14 @@
               {{/if}}
             </td>
           </tr>
+        {{else}}
+          <tr>
+            <td colspan="9" align="center">
+              {{if this.sessionQuery "No Results Found" "No Session Learning Materials Available"}}
```
Comment (stopfstedt) [inline addon/templates/components/course-materials.hbs]: same here, these messages need to be translatable.

File: addon/templates/components/course-materials.hbs (hunk 13 above)
Comment (djvoa12) [reply addon/templates/components/course-materials.hbs]: Fixed.

File: addon/templates/components/course-materials.hbs (hunk 14 above)
Comment (djvoa12) [reply addon/templates/components/course-materials.hbs]: Fixed.
